Faster collect_locations and playthrough generation

Zannick opened the pull request:
This implements a BFS simulation of a playthrough, tracking the regions accessed in each sphere and caching those regions and the exits to new regions that can't yet be taken, rather than use the can_reach inward DFS (inward because it starts at the goal location and searches backward until it finds an accessed region). This cuts generation time down nearly 50% depending on settings, and hopefully it's easier to understand. There may be more room for the Fill algorithm to take advantage of the new Playthrough caching, but the majority of time spent is in create_playthrough so that was what I was focusing on.

Experimental times on my laptop:
- Accessible, count=10:  33s -> 19s
- Accessible, worlds=5: 2m16s -> 1m10s
- Sanity (ocarina/egg/songs/cows/scrubs/shops/tokens/keys/maps+compasses): 5.5s -> 2.8s
- Sanity x10 with custom script to skip playthrough generation: 34s -> 18s

r0bd0g commented:
Does this work right?

r0bd0g commented:
" # This ensures the playthrough shows items being collected in the proper order."
Can you explain to me why this part is no longer necessary?

Zannick commented:
I misunderstood the intent of that section and assumed I had it covered with the other section. It looks like I did not. Fixed, thank you.

The playthroughs for my test seed are slightly different still due to the changes in iteration order within a sphere (eg. where to find Deku Sticks), but this should at least make them accurate. I don't see any differences in WotH.

Zannick commented:
Fixed having disabled locations show up in playthroughs by readding it to the accessibility check, but I think the list of disabled locations could be cached? Certainly it'd be easy to exclude exactly the locations listed in disabled_locations, but I'm not really sure I understand the DisableType stuff.

TestRunnerSRL commented:
So not sure if you're aware, but if you have a dictionary or set you intend to iterate over, then you need to make sure the keys properly implement hashing functions. They must have both an equals and a hash function and if two objects evaluate as equal, then the has must be equal. By default, the Obj class just uses the object's pointer for the hash/equals method. The problem is that this is not consistent for the same seeding, the ordering will change on each generation and is effectively random. There was a very similar bug with this a while back in the State class. You can try to resolve this by having implementing these functions, or by storing something that already has the implemented, like Strings. Which is why a lot of these structs have been storing lists/sets of strings instead of the actual obj they represent.

The other problem I have is that I'm not really a huge fan of how extra accessibility rules are generated. I'm worried about this becoming harder to maintain in the future. Would it be possible to call the original can_reach rules still? If you're just trying to ignore the parent reachability, maybe add some kind of parameter to ignore the parent can_reach. Something like this already exists for locations since that is how beatable_only works of the most part.

Zannick commented:
Thanks for the feedback.

I'd prefer to keep the actual objects themselves in the sets to avoid constantly needing to map get_location, especially since I'm keeping all the worlds' locations in the same set instead of splitting them up. So I'll write some hashing functions, maybe something like hash(name + world + seed).

As for the other part, yeah, sorry, sloppy of me to use the access rules directly like that. I will add a noparent parameter to Location.can_reach and Entrance.can_reach (I don't think I want to mix State.can_reach into this).

Zannick commented:
Ah, python's hashing semantics include hash randomization that is intentionally different on each run, so order will always be effectively random for sets. This is affecting the order of WotH locations and locations in a playthrough sphere... and hints. :( (Likely there may also be cases where it affects what's in a sphere.)

This is apparently not the case for dicts which are maintaining insertion order, so I'm going to have to replace the sets by lists and dicts and redo some of the set logic.

Zannick commented:
I've replaced the sets that were causing the nondeterministic spoiler output, and simplified the access checks to be noparent calls to can_reach. For my worlds=5 test case, this is a pretty minor performance hit (down to 1m15s-ish) within the margin of error so I'm happy with it. I've also added MS to BDoT -> ToT which makes it required. (Otherwise, the algorithm would never need MS as it bypasses can_become_adult for speed reasons.)
